tests/src/test_cases/widgets/test_canvas.c

```diff
@@ -0,0 +1,64 @@
+#if LV_BUILD_TEST
+#include "../lvgl.h"
+
+#include "unity/unity.h"
+
+static lv_obj_t * g_screen_active;
+
+void setUp(void)
+{
+    g_screen_active = lv_screen_active();
+}
+
+void tearDown(void)
+{
+    lv_obj_clean(g_screen_active);
+}
+
+static void draw_event_cb(lv_event_t * e)
+{
+    int * draw_counter = lv_event_get_user_data(e);
+    (*draw_counter)++;
+}
+void test_canvas_functions_invalidate(void)
+{
+    lv_obj_t * canvas = lv_canvas_create(g_screen_active);
+    int draw_counter = 0;
+    lv_obj_add_event_cb(canvas, draw_event_cb, LV_EVENT_DRAW_MAIN, &draw_counter);
+    lv_refr_now(NULL);
+    TEST_ASSERT(draw_counter == 0);
+
+    LV_DRAW_BUF_DEFINE(draw_buf, 100, 100, LV_COLOR_FORMAT_NATIVE);
+    lv_canvas_set_draw_buf(canvas, &draw_buf);
+    lv_refr_now(NULL);
+    TEST_ASSERT(draw_counter == 1);
+
+    lv_canvas_set_px(canvas, 0, 0, lv_color_black(), LV_OPA_COVER);
+    lv_refr_now(NULL);
+    TEST_ASSERT(draw_counter == 2);
+
+    lv_canvas_fill_bg(canvas, lv_color_black(), LV_OPA_COVER);
+    lv_refr_now(NULL);
+    TEST_ASSERT(draw_counter == 3);
+
+    {
```

Comment:
Why the `{` is needed?

Reply:
Removed. It's not needed. I just wanted to limit the scope of `layer` and `line_dsc`.